src/dshould/ShouldType.d

```diff
@@ -0,0 +1,218 @@
+module dshould.ShouldType;
+
+import std.algorithm : map;
+import std.format : format;
+import std.range : iota;
+import std.string : empty, join;
+import std.traits : TemplateArgsOf;
+public import std.traits : isInstanceOf;
+import std.typecons : Tuple;
+
+public auto should(T)(lazy T got) pure
+{
+    T get() pure @safe
+    {
+        return got;
+    }
+
+    return ShouldType!(typeof(&get))(&get);
+}
+
+public struct ShouldType(G, string[] Words = [])
+{
+    import std.algorithm : canFind;
+
+    public G got;
+
+    private int* refs_ = null;
+
+    public auto addWord(string Word)()
+    {
+        return ShouldType!(G, Words ~ Word)(this.got, this.refs);
+    }
+
+    private this(G got) { this.got = got; this.refs = 1; }
+
+    public this(G got, ref int refs) @trusted
+    in
+    {
+        assert(refs != CHAIN_TERMINATED, "don't copy Should that's been terminated");
+    }
+    body
+    {
+        this.got = got;
+        this.refs_ = &refs;
+        this.refs++;
+    }
+
+    this(this) @trusted
+    {
+        assert(this.refs != CHAIN_TERMINATED);
+
+        this.refs++;
+    }
+
+    ~this() @trusted
+    {
+        this.refs--;
+        assert(this.refs > 0, "unterminated should-chain!");
+    }
+
+    public enum gencheck(string TestString, string[] ArgNames) = gencheckImpl(TestString, ArgNames);
+
+    private static string gencheckImpl(string testString, string[] argNames)
+    {
+        auto args = argNames.join(", ");
+        auto argStrings = argNames.length.iota.map!(i => format!`"%s"`(argNames[i])).join(", ");
```

Comment:
auto argStrings = format!"%(%s, %)"(argNames);

Reply:
Note the inner quotes in the ``.

Reply:
Admittedly I have no idea what the point of the iota is.

Reply:
Oh wait, you're right.